mavlink_vehicles/mavlink_vehicles.cc

```diff
@@ -708,8 +759,24 @@ void mav_vehicle::set_mode(mode m, int timeout)
         // a custom mode instead. GUIDED mode is defined as 4.
         mavlink_message_t mav_msg;
         mavlink_set_mode_t mav_cmd_set_mode;
-        mav_cmd_set_mode.base_mode = MAV_MODE_FLAG_CUSTOM_MODE_ENABLED;
-        mav_cmd_set_mode.custom_mode = 4; // mode::GUIDED == 4
+        mav_cmd_set_mode.target_system = defaults::target_system_id;
+
+        switch (this->flight_stack_type) {
+        case firmware_type::PX4:
+            union px4_custom_mode px4_mode;
+            px4_mode.data = 0;
+            px4_mode.main_mode = PX4_CUSTOM_MAIN_MODE_OFFBOARD;
+            px4_mode.sub_mode = 0;
```

Comment:
PX4 doesn't have a GUIDED mode. In order to achieve a similar behavior we need to put PX4 in OFFBOARD mode by sending a specific custom_mode composed by the main mode PX4_CUSTOM_MAIN_MODE_OFFBOARD and an empty sub_mode.